### ISSUE TITLE
Validation without a FORM

### ISSUE THREAD

**Issue by cowwoc** [jquery-validation/jquery-validation#330]
There is high demand for validating fields on the client side without ever submitting them. For example, imagine a client-side mortgage calculator. Nothing ever returns to the server yet we need to validate the fields.

See also http://stackoverflow.com/questions/489722/jquery-ui-dialog-validation-without-using-form-tags

**Comment by dong-liu** [comment]
Please consider to extend the validation to the div tags. With modern dynamic web pages, a lot of people use div tags to make complicated pages, not just form. Please consider, thank you!

**Comment by didsun** [comment]
Yes, please. I need too.

**Comment by nhouse** [comment]
I have run into this issue as well as I need to validate inputs outside of a form. Putting a form around these inputs would result in a nested form, which is not allowed.

I haven't looked at the source code of this plugin, but I don't understand why a `<form>` tag is necessary for it to work. It seems like the plugin should just find the inputs by using jQuery's `.children()` or similar method, so why does there need to be a form around the inputs?

**Comment by staabm** [comment]
Atm `<form>` it the only valid validation-root because of some technical benefits we get from it.

Using a form it is very easy to find alle its contained elements, but also starting from a input-element the DOM provides properties to find its form.
This would be something we would need to work around when using divs.

Furthermore using a div we would no longer be able to use a delegated submit event on the form/container. We would need to simulate this event then.

Another point is that for semantic reasons your input fields should be part of a form. When using divs this a browser without javascript would not be able to submit your input fields (because of the missing form). 

To sum this up, I think using divs will have a lot of drawbacks. I think we should discuss before why at the first place you do not want to use a form. What are arguments in using div instead of form?

**Comment by nhouse** [comment]
> Furthermore using a div we would no longer be able to use a delegated submit event on the form/container. We would need to simulate this event then.

That seems like the biggest downside to me, although perhaps there's a way to retain this functionality for when a form is used? Worst case scenario, if a dev wants the enter key to submit the form (which is what I assume you're talking about, since a button can just have an onclick attribute), the container for the inputs (or the inputs themselves) can be given a class and a keydown handler can be assigned that submits the inputs when the enter key is pressed.

> Another point is that for semantic reasons your input fields should be part of a form. 

The issue is that nested forms are not supported, and often times web devs need to structure web pages in a way that would result in nested forms were form tags used. In my case, I'm working on an ecommerce platform and I need to logically separate groups of inputs on the pages for validation and submission, but having the inputs inside form tags would result in nested forms. No way around it.

> When using divs this a browser without javascript would not be able to submit your input fields (because of the missing form).

I don't think this matters anymore. Up until around a year ago I made sure all of my sites worked reasonably well without JavaScript, but the reality is the vast majority of people have browsers with JavaScript support. 

I tried Googling for the percent of users without JS support and stumbled across this answer which I think is really good: http://stackoverflow.com/a/17073999/1101095

>Give the basic info, with a clear route for how to go further - update your browser!
>
>I think sacrificing functionality for 99% of users to accommodate 1% is sheer bloody mindedness.
>
>Sure, it is possible to allow for non-javascript enabled content for every aspect of a website, plus provide the optimal experience - but the budget is going to sky-rocket for the build.
>
>There's some seriously awesome stuff going down with Javascript which actually makes sites far more accessible! - where do we draw the line here?
>
>"Sorry, your computer is too old and slow to render this website." OR
>
>"Sorry, 99.9% of the planet, we've presented you with a sub-optimal 1993 experience because 0.1% of you have outdated tech"
>
>I don't buy the '1%' is important argument - if someone Really wants to access a website, they'll find a way - plus those instances where companies are forced to use, say, ie7 with javascript turned off - heck, they probably aren't allowed to browse anything but the corporate intranet anyway!
>
>Time to get off this dumb old idea you need to have a non-javascript option for everything on your website, it really is an outdated concept.

**Comment by staabm** [comment]
> That seems like the biggest downside to me, although perhaps there's a way to retain this functionality for when a form is used? Worst case scenario, if a dev wants the enter key to submit the form (which is what I assume you're talking about, since a button can just have an onclick attribute), the container for the inputs (or the inputs themselves) can be given a class and a keydown handler can be assigned that submits the inputs when the enter key is pressed.

The worst thing is that we need to simulate all the functionality, the browser already has builtin. Some browser offer even more features like this basic ones which we might need to simulate then.

> The issue is that nested forms are not supported, and often times web devs need to structure web pages in a way that would result in nested forms were form tags used. In my case, I'm working on an ecommerce platform and I need to logically separate groups of inputs on the pages for validation and submission, but having the inputs inside form tags would result in nested forms. No way around it.

Are `fieldset`s not an option for you? 
Another negative point because you dont adhere to semantic html is, that other tools can no  longer built  easily on top of your html, e.g. Password autocomplete.

> I tried Googling for the percent of users without JS support and stumbled across this answer which I think is really good: http://stackoverflow.com/a/17073999/1101095

Starting google will also reveal that there are a lot more reasons then browsers not understanding JS.
mobile devices take a lot of time to retrieve JS scripts on low bandwidth sometimes. Sometimes retrievals fails completly which in turn renders your page useless.

**Comment by danicomas** [comment]
I think the same than @nhouse. Many problems appear when you want to validate different field groups (divs inside a div and validate only the panels shown). To do submits with postbacks is not a modern web app way.

**Comment by staabm** [comment]
> I think the same than @nhouse. Many problems appear when you want to validate different field groups (divs inside a div and validate only the panels shown). To do submits with postbacks is not a modern web app way.

Every good js book propagate progressive enhancement, so programming that everything works without js.

Could someone please give more concrete examples of problems you encouter. What I read until now was nothing concrete but people argumenting with their lazyness.
Please dont take this as offence, I just try to describe what we need to progress further on this topic.

**Comment by danicomas** [comment]
@jzaefferer approved this PR instead than my PR but the dev after do the first commit didn't correct the style and validation errors. If anyone want to do the changes is this: https://github.com/jzaefferer/jquery-validation/pull/1161

**Comment by danicomas** [comment]
The problem is this:
```html
<div>
(*1)
customer fields (name, comments,...)
 <div>
(*2)
locations of the customer with other fields to add, update and delete and a small grid.
 </div>
CRUD buttons of customer
</div>

<div>
customers main grid
</div>
```

(*1 and *2 should share the same div but not the same validation! if you use the the library now you need to enable or disable the validators that you want every time. Of course, it could be improved, isn't it? @staabm)

**Comment by danicomas** [comment]
In some frameworks use validation groups in each field and in submit button too but I think that div approach is better.

**Comment by jzaefferer** [comment]
@danicomas could you put together a testpage based on your outline above? Something we can use to explore alternatives and test solutions.

**Comment by RJCuthbertson** [comment]
What is the state of the implementation of this feature request? Is it being implemented? Was it rejected? Is it sitting on a // TODO list somewhere?

**Comment by staabm** [comment]
I personally dont think it is required there not on the todo list. A well crafted PR with unit tests and no BC break is welcome though

**Comment by churiart** [comment]
Answer to this issue is really appreciated.

**Comment by staabm** [comment]
> Answer to this issue is really appreciated.

see the comment above yours.

**Comment by churiart** [comment]
I'm talking about a workaround. @staabm

**Comment by staabm** [comment]
@churiart feel free to provide a PR with unit tests and we will see. I dont have a use-case in my applications for this sort of validation, therefore dont expect me to implement it.

**Comment by churiart** [comment]
@staabm  Well let me elaborate on this a bit. Validation plugin have some conflicts with [XEditable](https://github.com/vitalets/x-editable) plugin which provides simple inline edits.

A use case could be if you have a form wizard that has multiple tabs in it, all should wrapped within a form element to work. In final tab, most of the time, there is a review table of what information you have provided. So If an XEditable object is initialized on this review list, it will create an input on clicking per particular field to enable editing its value. On DOM getting these changes, Validation plugin will throw an error: `Uncaught TypeError: Cannot read property 'settings' of undefined`.

But If I had this ability to select all tabs except final tab instead of selecting a form, it wouldn't happen.

**Comment by Fullmight** [comment]
It's really unfortunate this issue has been dismissed for so long. It's starting to look like I'm going to have to write my on validation from scratch because of the common dependency on <form>, but for example at the company I work at, it is impossible to wrap our forms in form tags.

**Comment by FernandoTBarros** [comment]
I don't understand why this issue is closed and nothing was made so far. Anyone knows any workaround or another plugins that work with divs?

**Comment by alancoughlin** [comment]
I am a designer working on a MS SharePoint 2013 project. The SharePoint developer tells me there is no way around the fact that on most pages SharePoint wraps the whole page in form tags and automatically strips out all other form tags. The fault is clearly with SharePoint for violating basic HTML standards and delivering malformed HTML - but that's MS for you. The problem is that I am insisting on our using jQuery Validation Plugin on all forms in the site because there is nothing that allows me to create a great user experience so easily - nothing! 

MS may have corrected their ways in future versions - I don't know. But I don't think there is much hope that MS will go back and fix SP 2013, so my only hope is that the request put forth in this thread is fulfilled. 

Please let me know if it has been done somewhere and provide a link.
